compiler-core/src/language_server/engine.rs

```diff
@@ -241,13 +245,33 @@ where
         self.respond(|this| {
             let params = params.text_document_position_params;
 
-            let (lines, found) = match this.node_at_position(&params) {
+            let (lines, byte_index, found) = match this.node_at_position(&params) {
                 Some(value) => value,
                 None => return Ok(None),
             };
 
             Ok(match found {
                 Located::Statement(_) => None, // TODO: hover for statement
+                Located::ModuleStatement(Definition::Function(fun)) => {
+                    if !fun.location.contains(byte_index) {
+                        // `fun.location` is just the head of the function, so this ensures that
+                        // hovering over the body doesn't produce any results.
+                        None
+                    } else {
+                        if let Some(arg) = fun
+                            .arguments
+                            .iter()
+                            .find(|arg| arg.location.contains(byte_index))
+                        {
+                            Some(hover_for_function_argument(arg, lines))
+                        } else {
+                            Some(hover_for_function_head(fun, lines))
+                        }
+                    }
```

Comment:
This finding logic should go in the `node_at_position` functions with the other finding logic that deals with byte indexes and AST. This function here doesn't perform any finding logic, it just maps from a found node to a hover response function.

Sorry, I was unclear previously!

Reply:
That sounds like the best way to do this, but I wanted to check if my understanding/approach sounds reasonable before I try to do stuff (also this is my first open-source contribution of non-trivial size, so this is a bit scary).

Does this sound right:

`node_at_position` should be capable of finding the function head and each function arg. The main type `node_at_position` returns is `Located`, and I don't think there's a way to represent a function argument in a `Located`. So I think I'd add an entry to `Located` called `Arg(&'a TypedArg)`. Then in [`TypedDefinition::find_node`](https://github.com/nino/gleam/blob/1149636aa3112abb2f7701506396e588abee024a/compiler-core/src/ast.rs#L548-L560), I'd try to return a `Located::Arg` if possible, and a `Located::ModuleStatement` if not?

(And also I'll look at how to write tests for the new behavior.)

Reply:
Yup! That's perfect

Reply:
Why is there still some location logic in this function? What is the case in which it is needed?

Reply:
The remaining logic is to prevent the function signature coming up when you hover leading/trailing whitespace or comments in the function body. It's essentially undoing the [`.full_location()` logic in `ast.rs`](https://github.com/nino/gleam/blob/051c6b04191681ff98f4db9786c3fd4bf6d4c685/compiler-core/src/ast.rs#L561-L567).